e2e/e2e_test.go

```diff
@@ -0,0 +1,206 @@
+	sb.StartManager(t, controller.InitClusterReconcilerWithLogger(testLog))
+
+	b := testutil.NewBuilder("crdb").WithNodeCount(1).WithEmptyDirDataStore()
+
+	create := Step{
+		name: "creates 1-node insecure cluster",
+		test: func(t *testing.T) {
+			require.NoError(t, sb.Create(b.Cr()))
+
+			requireClusterToBeReadyEventually(t, sb, b)
+
+			state, err := sb.Diff()
+			require.NoError(t, err)
+
+			expected := testutil.ReadOrUpdateGoldenFile(t, state, *updateOpt)
+
+			testutil.AssertDiff(t, expected, state)
+		},
+	}
+
+	steps := Steps{create}
+
+	steps.Run(t)
+}
+
+func TestCreatesSecureClusterWithGeneratedCert(t *testing.T) {
+	if testing.Short() {
+		t.Skip("skipping test in short mode.")
+	}
+
+	testLog := zapr.NewLogger(zaptest.NewLogger(t))
+
+	actor.Log = testLog
+
+	sb := testenv.NewDiffingSandbox(t, env)
+	sb.StartManager(t, controller.InitClusterReconcilerWithLogger(testLog))
+
+	b := testutil.NewBuilder("crdb").WithNodeCount(1).WithTLS().WithEmptyDirDataStore()
+
+	create := Step{
+		name: "creates 1-node secure cluster",
+		test: func(t *testing.T) {
+			require.NoError(t, sb.Create(b.Cr()))
+
+			requireClusterToBeReadyEventually(t, sb, b)
+
+			state, err := sb.Diff()
+			require.NoError(t, err)
+
+			expected := testutil.ReadOrUpdateGoldenFile(t, state, *updateOpt)
+
+			testutil.AssertDiff(t, expected, state)
+		},
+	}
+
+	steps := Steps{create}
+
+	steps.Run(t)
+}
+
+func TestUpgradesMinorVersion(t *testing.T) {
+	if testing.Short() {
+		t.Skip("skipping test in short mode.")
+	}
+
+	testLog := zapr.NewLogger(zaptest.NewLogger(t))
+
+	actor.Log = testLog
+
+	sb := testenv.NewDiffingSandbox(t, env)
+	sb.StartManager(t, controller.InitClusterReconcilerWithLogger(testLog))
+
+	builder := testutil.NewBuilder("crdb").WithNodeCount(1).WithTLS().
+		WithImage("cockroachdb/cockroach:v19.2.5").
+		WithPVDataStore("1Gi", "standard" /* default storage class in KIND */)
+
+	steps := Steps{
+		{
+			name: "creates a 1-node secure cluster",
+			test: func(t *testing.T) {
+				require.NoError(t, sb.Create(builder.Cr()))
+
+				requireClusterToBeReadyEventually(t, sb, builder)
+			},
+		},
+		{
+			name: "upgrades the cluster to the next minor version",
+			test: func(t *testing.T) {
+				current := builder.Cr()
+				require.NoError(t, sb.Get(current))
+
+				current.Spec.Image = "cockroachdb/cockroach:v19.2.6"
+				require.NoError(t, sb.Update(current))
+
+				requireClusterToBeReadyEventually(t, sb, builder)
+				requireDbContainersToUseImage(t, sb, current)
+			},
+		},
+	}
+
+	steps.Run(t)
+}
+
+func TestUpgradesMajorVersion(t *testing.T) {
+	if testing.Short() {
+		t.Skip("skipping test in short mode.")
+	}
+
+	testLog := zapr.NewLogger(zaptest.NewLogger(t))
+
+	actor.Log = testLog
+
+	sb := testenv.NewDiffingSandbox(t, env)
+	sb.StartManager(t, controller.InitClusterReconcilerWithLogger(testLog))
+
+	builder := testutil.NewBuilder("crdb").WithNodeCount(1).WithTLS().
+		WithImage("cockroachdb/cockroach:v19.2.6").
```

Comment:
Can you test an upgrade from `v19.1.4` to `v19.2.1` or something also? Is a bit different in terms of the change in version tag but also a major release according to our definitions at CRL.

Reply:
I actually thought about having 19.2 as the minimal supported version by the operator, but if you really think we need such test, I could add it.

Reply:
I've added an additional e2e test.